src/librustc_resolve/lib.rs

```diff
@@ -797,36 +787,103 @@ bitflags! {
     }
 }
 
-// Records a possibly-private type definition.
+// Records a possibly-private definition.
 #[derive(Clone,Debug)]
-struct TypeNsDef {
+struct NsDef {
     modifiers: DefModifiers, // see note in ImportResolution about how to use this
-    module_def: Option<Rc<Module>>,
-    type_def: Option<Def>,
-    type_span: Option<Span>
+    def: Option<Def>,
+    module: Option<Rc<Module>>,
```

Comment:
If both `def` and `module` are `Some` it means that the `NsDef` represents both a `struct` definition and a `mod` definition. This can happen if and only if either a tuple `struct` and a `mod` have the same name or a braced `struct` and a `mod` have the same name and the `struct` is defined first (see [this issue](https://github.com/rust-lang/rust/issues/21546)). This [pull request](https://github.com/rust-lang/rust/pull/26421) adds a warning for these cases.

I believe `def` takes precedence since the rest of the code uses `NameBinding::def()` and `NameBinding::module()` instead of the `def` and `module` fields so if both fields were `Some` and `module` took precedence there would be no way to see the `def`. I'm not entirely sure though -- I wrote `NsDef::def()` so that its semantics matched the old `NameBindings::def_for_namespace()`, which has the `type_def` field of a `TypeNsDef` taking precedence when the namespace is `TypeNS`.

Reading the comments of the pull request I linked in more detail, it looks like it was decided that there would be one warning cycle before the warnings become errors. Once they are errors, I could refactor the `def` and `module` fields into something isomorphic to `Result<Def, Rc<Module>>` very easily.

I checked that the warnings are currently in the beta branch. Does this means they are ready to be turned into errors now? If so I could do that and refactor the fields in this PR.